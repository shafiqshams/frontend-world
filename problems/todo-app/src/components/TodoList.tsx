import { useState } from 'react'
import { Todo } from '../types/todo';
import TodoItem from './TodoItem';


const TodoList = () => {

   const [todos, setTodos] = useState<Todo[]>([]);
   const [input, setInput] = useState<string>('');


   const handleAddTodo = () => {
      if (!input.trim()) return;

      const newTodo: Todo = {
         id: Date.now(),
         title: input,
         completed: false
      }

      setTodos([...todos, newTodo]);
      setInput('');
   }

   const handleToggleTodo = (todoId: number) => {
      const updatedTodos: Todo[] = todos.map(todo => todo.id === todoId ? { ...todo, completed: !todo.completed } : todo)
      setTodos(updatedTodos)
   }

   const handleDeleteTodo = (todoId: number) => {
      const updatedTodos: Todo[] = todos.filter(todo => todo.id !== todoId);
      setTodos(updatedTodos);
   }

   return (
      <>
         <div>Todo App</div>

         <input
            value={input}
            onChange={(e) => setInput(e.target.value)}
            placeholder='Write here...'
            type='text'
            data-testid='todo-input'
         />

         <button
            data-testid='add-btn'
            onClick={handleAddTodo}>Add todo</button>

         {todos.map((todo) =>
            <TodoItem
               key={todo.id}
               todo={todo}
               deleteTodo={handleDeleteTodo}
               toggleTodo={handleToggleTodo}
            />
         )}
      </>
   )
}

export default TodoList