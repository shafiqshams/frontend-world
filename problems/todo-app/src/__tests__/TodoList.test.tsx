import TodoList from "../components/TodoList";
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('TodoList', () => {
   test('renders input and button', () => {
      //Arrange
      const { asFragment, getByTestId } = render(<TodoList />);
      const button = getByTestId('add-btn');
      const input = getByTestId('todo-input')
      //Act
      // No action - only rendering

      //Assert
      expect(asFragment()).toMatchSnapshot();
      // expect(button).toBeInTheDocument();
      expect(button).toBeDefined();
      expect(input).toBeDefined();
      // expect(input).toBeInTheDocument();
   });

   test('adds new todo to the list when btn is pressed', () => {

      //Arrange
      const { getByTestId, getByText } = render(<TodoList />);
      const input = getByTestId('todo-input');
      const btn = getByTestId('add-btn');

      //Act
      fireEvent.change(input, { target: { value: 'Test todo' } });
      fireEvent.click(btn);

      //Assert
      expect(getByText('Test todo')).toBeInTheDocument();
   })

   //Integration test
   test('marks todo as completed and applies line-through when clicked', () => {

      //Arrange
      const { getByTestId, getByText } = render(<TodoList />)
      const input = getByTestId('todo-input');
      const btn = getByTestId('add-btn');

      //Act: add a todo
      fireEvent.change(input, { target: { value: 'Test todo' } })
      fireEvent.click(btn);

      //Get the text of newly added todo
      const todoText = getByText('Test todo');

      //Assert: its rendered properly
      expect(todoText).toBeInTheDocument();
      expect(todoText).not.toHaveStyle('text-decoration: line-through');

      //Act2: toggle the todo by clicking it
      fireEvent.click(todoText);

      //Assert it should now be marked completed with line-through
      expect(todoText).toHaveStyle('text-decoration: line-through');
   })

});