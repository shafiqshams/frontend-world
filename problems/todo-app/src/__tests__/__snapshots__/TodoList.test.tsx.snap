// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders input and button 1`] = `
<DocumentFragment>
  <div>
    Todo App
  </div>
  <input
    data-testid="todo-input"
    placeholder="Write here..."
    type="text"
    value=""
  />
  <button
    data-testid="add-btn"
  >
    Add todo
  </button>
</DocumentFragment>
`;
